server/src/main/java/com/thoughtworks/go/server/service/ScheduleService.java

```diff
@@ -251,8 +241,43 @@ private boolean canSchedule(PipelineConfig pipelineConfig) {
         return schedulingChecker.canAutoTriggerConsumer(pipelineConfig);
     }
 
-    public Stage rerunStage(String pipelineName, String counterOrLabel, String stageName) {
-        return lockAndRerunStage(pipelineName, counterOrLabel, stageName, new NewStageInstanceCreator(goConfigService), new ExceptioningErrorHandler());
+    public boolean rerunStage(Pipeline pipeline, StageConfig stageConfig, String approvedBy) {
+        internalRerun(pipeline, CaseInsensitiveString.str(stageConfig.name()), approvedBy, new NewStageInstanceCreator(goConfigService), new ExceptioningErrorHandler());
+        return true;
+    }
+
+    public Stage rerunStage(String pipelineName, String pipelineCounter, String stageName) {
+        return rerunStage(pipelineName, pipelineCounter, stageName, new ExceptioningErrorHandler());
+    }
+
+    public Stage rerunStage(String pipelineName, String pipelineCounter, String stageName, ErrorConditionHandler errorHandler) {
+        return lockAndRerunStage(pipelineName, pipelineCounter, stageName, new NewStageInstanceCreator(goConfigService), errorHandler);
+    }
+
+    /**
+     * Top-level operation only; consumes exceptions
+     */
+    public Stage rerunStage(String pipelineName, String pipelineCounter, String stageName, HttpOperationResult result) {
+        String identifier = StringUtils.join(Arrays.asList(pipelineName, pipelineCounter, stageName), "/");
+        HealthStateType healthStateType = HealthStateType.general(HealthStateScope.forStage(pipelineName, stageName));
+        Stage stage = null;
+
+        try {
+            stage = rerunStage(pipelineName, pipelineCounter, stageName, new ResultUpdatingErrorHandler(result));
+
+            if (result.isSuccess()) { // this check is for clarity, but is not needed as failures result in RunTimeExceptions
+                result.accepted(String.format("Request to schedule stage %s accepted", identifier), "", healthStateType);
+            }
+        } catch (RuntimeException e) {
+            String message = String.format(
+                    "Stage rerun request for stage [%s] could not be completed " +
+                            "because of an unexpected failure. Cause: %s", identifier,
+                    e.getMessage()
+            );
+            LOGGER.error(message, e);
+            result.internalServerError(message, healthStateType);
```

Comment:
We're catching all the RuntimeExceptions and returning `500`. Is there something we can do about it? I mean rather than throwing/catching runtime exceptions, can we create specific exceptions that might occur while triggering a stage?

Reply:
Also, didn't see any unit and the integration tests for this code.

Reply:
> We're catching all the RuntimeExceptions and returning `500`. Is there something we can do about it? I mean rather than throwing/catching runtime exceptions, can we create specific exceptions that might occur while triggering a stage?

I'm going to apply this patch and merge the PR

```patch
diff --git a/server/src/main/java/com/thoughtworks/go/server/service/ScheduleService.java b/server/src/main/java/com/thoughtworks/go/server/service/ScheduleService.java
index 48b8d12bc..df0cea198 100644
--- a/server/src/main/java/com/thoughtworks/go/server/service/ScheduleService.java
+++ b/server/src/main/java/com/thoughtworks/go/server/service/ScheduleService.java
@@ -269,13 +269,17 @@ public class ScheduleService {
                 result.accepted(String.format("Request to schedule stage %s accepted", identifier), "", healthStateType);
             }
         } catch (RuntimeException e) {
-            String message = String.format(
-                    "Stage rerun request for stage [%s] could not be completed " +
-                            "because of an unexpected failure. Cause: %s", identifier,
-                    e.getMessage()
-            );
-            LOGGER.error(message, e);
-            result.internalServerError(message, healthStateType);
+            // if the result is successful, but there's still an exception, treat it as a 500
+            // else the result is assumed to contain the right status code and error message
+            if (result.isSuccess()) {
+                String message = String.format(
+                        "Stage rerun request for stage [%s] could not be completed " +
+                                "because of an unexpected failure. Cause: %s", identifier,
+                        e.getMessage()
+                );
+                LOGGER.error(message, e);
+                result.internalServerError(message, healthStateType);
+            }
         }
         return stage;
     }
```